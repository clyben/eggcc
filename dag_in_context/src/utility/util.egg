(function ListExpr-length (ListExpr) i64)
(function ListExpr-ith (ListExpr i64) Expr :unextractable)
(function ListExpr-suffix (ListExpr i64) ListExpr :unextractable)
(function Append (ListExpr Expr) ListExpr :unextractable)

(rule ((Switch pred inputs branch)) ((union (ListExpr-suffix branch 0) branch)) :ruleset always-run)

(rule ((= (ListExpr-suffix top n) (Cons hd tl)))
    ((union (ListExpr-ith top n) hd)
     (union (ListExpr-suffix top (+ n 1)) tl)) :ruleset always-run)

(rule ((= (ListExpr-suffix list n) (Nil)))
    ((set (ListExpr-length list) n)) :ruleset always-run)

(rewrite (Append (Cons a b) e)
   (Cons a (Append b e))
   :ruleset always-run)
(rewrite (Append (Nil) e)
   (Cons e (Nil))
   :ruleset always-run)

(function tuple-length (Expr) i64 :unextractable)

(rule ((HasType expr (TupleT tl))
       (= len (TypeList-length tl)))
      ((set (tuple-length expr) len)) :ruleset always-run)

;; Create a Get for every index, and rewrite it to see through Concat
(rule ((Single expr)) ((union (Get (Single expr) 0) expr)) :ruleset always-run)
;; initial get
(rule ((> (tuple-length tuple) 0))
      ((Get tuple 0))
      :ruleset always-run)
;; next get
(rule ((= len (tuple-length tuple))
       (= ith (Get tuple i))
       (< (+ i 1) len)
       )
       ((Get tuple (+ 1 i)))
       :ruleset always-run)

;; descend left
(rule ((Get (Concat expr1 expr2) i)
       (= (tuple-length expr1) len1)
       (< i len1))
      ((union (Get (Concat expr1 expr2) i)
              (Get expr1 i)))
       :ruleset always-run)
;; descend right
(rule ((Get (Concat expr1 expr2) i)
       (= (tuple-length expr1) len1)
       (>= i len1))
      ((union (Get (Concat expr1 expr2) i)
              (Get expr2 (- i len1))))
       :ruleset always-run)


(relation IsLeaf (Expr))
(rule ((Arg ty)) ((IsLeaf (Arg ty))) :ruleset always-run)
(rule ((Const val ty)) ((IsLeaf (Const val ty))) :ruleset always-run)
(rule ((Empty ty)) ((IsLeaf (Empty ty))) :ruleset always-run)

(rewrite (Concat (Concat a b) c)
         (Concat a (Concat b c))
         :ruleset always-run)



;; This Expr size actually do the same thing as cost function.
(function Expr-size (Expr) i64 :unextractable :merge (min old new) )
(function ListExpr-size (ListExpr) i64 :unextractable :merge (min old new))

(rule ((= expr (Function name tyin tyout out)) 
       (= sum (Expr-size out))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Const n ty))) 
      ((set (Expr-size expr) 1))  :ruleset always-run)

(rule ((= expr (Bop op x y)) 
       (= sum (+ (Expr-size y) (Expr-size x)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Uop op x)) 
       (= sum (Expr-size x))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Get tup i)) 
       (= sum (Expr-size tup))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Concat x y)) 
       (= sum (+ (Expr-size y) (Expr-size x)))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Single x)) 
       (= sum (Expr-size x))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Switch pred branches)) 
       (= sum (+ (ListExpr-size branches) (Expr-size pred)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (If pred then else)) 
       (= sum (+ (Expr-size else) (+ (Expr-size then) (Expr-size pred))))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (DoWhile in pred-and-output)) 
       (= sum (+ (Expr-size pred-and-output) (Expr-size in)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Arg ty))) 
      ((set (Expr-size expr) 1)) :ruleset always-run)

(rule ((= expr (Call func arg)) 
       (= sum (Expr-size arg))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((Empty ty)) ((set (Expr-size (Empty ty)) 0))  :ruleset always-run)

(rule ((= expr (Cons hd tl)) 
       (= sum (+ (ListExpr-size tl) (Expr-size hd)))) 
      ((set (ListExpr-size expr) sum)) :ruleset always-run)

(rule ((Nil)) 
      ((set (ListExpr-size (Nil)) 0))  :ruleset always-run)

(rule ((= expr (Alloc id e state ty)) ;; do state edge's expr should be counted?
        (= sum (Expr-size e))) 
        ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (InContext assumption e)) 
       (= sum (Expr-size e))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)


;;                tuple start len
(function SubTuple (Expr i64 i64) Expr :unextractable)
(rewrite (SubTuple expr x 1)
         (Single (Get expr x)) :ruleset always-run)
(rule ((= f (SubTuple expr a b)) (> b 1))
      ((union f 
         (Concat (Single (Get expr a))
                 (SubTuple expr (+ a 1) (- b 1))))) :ruleset always-run)
