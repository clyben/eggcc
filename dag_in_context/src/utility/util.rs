#[cfg(test)]
use crate::{egglog_test, interpreter::Value};

#[test]
fn test_list_util() -> crate::Result {
    use crate::ast::emptyt;
    let emptyt = emptyt();
    let build = format!(
        "
		(let list (Cons (Const (Int 0) {emptyt} (NoContext))
                  (Cons (Const (Int 1) {emptyt} (NoContext))
                  (Cons (Const (Int 2) {emptyt} (NoContext))
                  (Cons (Const (Int 3) {emptyt} (NoContext))
                  (Cons (Const (Int 4) {emptyt} (NoContext)) (Nil)))))))
		(let expr (Switch (Const (Int 1) {emptyt} (NoContext)) (Empty {emptyt} (NoContext)) list))
	"
    );
    let check = format!(
        "
		(check (= (ListExpr-ith list 1) (Const (Int 1) {emptyt} (NoContext))))
        (check (= (ListExpr-ith list 4) (Const (Int 4) {emptyt} (NoContext))))
        (check (= (ListExpr-length list) 5))
	"
    );
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn append_test() -> crate::Result {
    use crate::ast::emptyt;
    let emptyt = emptyt();
    let build = format!(
        "
        (let appended
            (Append
                (Cons (Const (Int 0) {emptyt} (NoContext)) (Cons (Const (Int 1) {emptyt} (NoContext)) (Nil)))
                (Const (Int 2) {emptyt} (NoContext))))
    "
    );

    let check = format!("
        (check (
            =
            (Cons (Const (Int 0) {emptyt} (NoContext)) (Cons (Const (Int 1) {emptyt} (NoContext)) (Cons (Const (Int 2) {emptyt} (NoContext)) (Nil))))
            appended
        ))
    ");
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn test_tuple_ith() -> crate::Result {
    use crate::ast::emptyt;
    let emptyt = emptyt();
    let build = format!(
        "
    (let tup (Concat
                  (Concat (Single (Const (Int 0) {emptyt} (NoContext))) (Single (Const (Int 1) {emptyt} (NoContext))))
                  (Concat (Single (Const (Int 2) {emptyt} (NoContext))) (Single (Const (Int 3) {emptyt} (NoContext))))))
    
    ;; with print
    (let tup2 (Concat 
                (Concat 
                    (Single (Bop (Print) (Const (Int 0) {emptyt} (NoContext)) (Arg (Base (StateT)) (NoContext))))
                    (Concat (Single (Const (Int 1) {emptyt} (NoContext))) 
                                (Single (Const (Int 2) {emptyt} (NoContext)))))
                (Single (Const (Int 3) {emptyt} (NoContext)))))
    "
    );

    let check = format!(
        "
    (check (= (Get tup 0) (Const (Int 0) {emptyt} (NoContext))))
    (check (= (Get tup 1) (Const (Int 1) {emptyt} (NoContext))))
    (check (= (Get tup 2) (Const (Int 2) {emptyt} (NoContext))))
    (check (= (Get tup 3) (Const (Int 3) {emptyt} (NoContext))))
    (check (= 4 (tuple-length tup)))
    (fail (check (Get tup 4)))

    (check (= (Get tup2 0) (Bop (Print) (Const (Int 0) {emptyt} (NoContext)) (Arg (Base (StateT)) (NoContext)))))
    (check (= (Get tup2 1) (Const (Int 1) {emptyt} (NoContext))))
    (check (= (Get tup2 2) (Const (Int 2) {emptyt} (NoContext))))
    (check (= (Get tup2 3) (Const (Int 3) {emptyt} (NoContext))))
    (check (= 4 (tuple-length tup2)))
    (fail (check (Get tup2 4)))
    "
    );
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn test_expr_size() -> crate::Result {
    use crate::ast::*;
    let pureloop = dowhile(
        single(int(1)),
        parallel!(
            less_than(getat(0), int(3)),
            get(switch!(int(2); parallel!(int(4), int(5))), 0)
        ),
    )
    .with_arg_types(emptyt(), tuplet!(intt()));
    let build: String = format!("(let loop {})", pureloop);

    let check = "(check (= 9 (Expr-size loop)))";
    egglog_test(
        build.as_str(),
        check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn test_sub_tuple() -> crate::Result {
    use crate::ast::*;
    let ty = tuplet!(intt(), intt(), intt());
    let arg = arg().with_arg_types(ty.clone(), ty.clone());
    let build = format!("(let expr (SubTuple {} 0 3))", arg);
    let out = concat(
        single(getat(0)),
        concat(single(getat(1)), single(getat(2))),
    )
    .with_arg_types(ty.clone(), ty);

    let check = format!("(check (= expr {}))", out);
    egglog_test(
        build.as_str(),
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}
