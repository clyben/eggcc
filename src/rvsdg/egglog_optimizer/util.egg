;; ####################################
;; implementation of PassThroughArguments
;; Creates a vec of arguments
;; (vec-of (Arg 0) (Arg 1) ...) with length i
(function PassThroughArguments (i64) VecOperand :unextractable)



;; (how many arguments to generate, vector so far)
(function PassThroughArgumentsHelper (i64 VecOperand) VecOperand :unextractable)

(rewrite (PassThroughArguments i)
  (PassThroughArgumentsHelper i (VO (vec-of)))
  :ruleset subst)

(rule ((= lhs (PassThroughArgumentsHelper i (VO rest)))
       (< (vec-length rest) i))
      ((union lhs
        (PassThroughArgumentsHelper i
            (VO (vec-push rest (Arg (vec-length rest)))))))
      :ruleset subst)

(rule ((= lhs (PassThroughArgumentsHelper i (VO rest)))
       (= (vec-length rest) i))
      ((union lhs (VO rest)))
      :ruleset subst)



;; Project each argument out of a body
(function BodyToVecOperand (i64 Body) VecOperand :unextractable)
;; current index, body length, body, and vector so far
(function BodyToVecOperandHelper (i64 i64 Body VecOperandBase) VecOperand :unextractable)

(rewrite (BodyToVecOperand body-len body)
         (BodyToVecOperandHelper 0 body-len body (vec-of)) :ruleset subst)
(rule
  ((= helper (BodyToVecOperandHelper index body-len body so-far))
   (< index body-len))
  ((union helper
          (BodyToVecOperandHelper (+ index 1) body-len body
                                  (vec-push so-far 
                                   (Project index body)))))
  :ruleset subst)

(rule
  ((= helper (BodyToVecOperandHelper index body-len body so-far))
   (= index body-len))
  ((union helper (VO so-far)))
  :ruleset subst)

;; Converts (vec n1 n2 ...) to (vec (Arg n1) (Arg n2) ...)
(function MapArgOverVec (VecInt) VecOperand :unextractable)
(function MapArgOverVecHelper (VecIntBase i64 VecOperandBase) VecOperand :unextractable)
(let vec-operand-empty (vec-pop (vec-of (Arg 0))))
(rewrite (MapArgOverVec (VI vec)) 
         (MapArgOverVecHelper vec 0 vec-operand-empty)
      :ruleset subst)
(rewrite (MapArgOverVecHelper vec i curr)
         (MapArgOverVecHelper vec (+ i 1) (vec-push curr (Arg (vec-get vec i))))
      :when ((< i (vec-length vec)))
      :ruleset subst)
(rewrite (MapArgOverVecHelper vec i curr)
         (VO curr)
      :when ((= i (vec-length vec)))
      :ruleset subst)

(function VecAppend (VecOperand VecOperand) VecOperand)
(rewrite (VecAppend (VO v1) (VO v2)) 
         (VO (vec-append v1 v2))
      :ruleset subst)

;; Converts (vec n1 n2 ...) to (map-of n1 (Arg offset) n2 (Arg (+ offset 1)) ...)
;;                           vec    offset
(function BuildSubstMapFrom (VecInt i64) MapIntOperand :unextractable)
(function BuildSubstMapFromHelper (VecIntBase i64 i64 MapIntOperandBase) MapIntOperand :unextractable)
(rewrite (BuildSubstMapFrom (VI vi) offset) 
         ;; because type checking for actions is broken, we 
         ;; have to make sure there is only one Map sort right now
         (BuildSubstMapFromHelper vi offset 0 (map-empty))
         :ruleset subst)
(rewrite (BuildSubstMapFromHelper vi offset i m)
         (BuildSubstMapFromHelper vi offset (+ i 1) (map-insert m (vec-get vi i) (Arg (+ i offset))))
      :when ((< i (vec-length vi)))
      :ruleset subst)
(rewrite (BuildSubstMapFromHelper vi offset i m)
         (MIO m)
      :when ((= i (vec-length vi)))
      :ruleset subst)
