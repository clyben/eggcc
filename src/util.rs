use crate::peg::rvsdg_to_peg;
use crate::Optimizer;
use hashbrown::HashMap;
use petgraph::dot::Dot;
use std::{
    ffi::OsStr,
    fmt::{Display, Formatter},
    io,
    path::PathBuf,
};

pub(crate) struct ListDisplay<'a, TS>(pub TS, pub &'a str);

impl<'a, TS> Display for ListDisplay<'a, TS>
where
    TS: Clone + IntoIterator,
    TS::Item: Display,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut did_something = false;
        for item in self.0.clone().into_iter() {
            if did_something {
                f.write_str(self.1)?;
            }
            Display::fmt(&item, f)?;
            did_something = true;
        }
        Ok(())
    }
}

/// Parse a string containing a bril program (in text format) into a Program.
///
/// This function is intended for use in tests and in ad-hoc debugging.
#[allow(unused)]
pub(crate) fn parse_from_string(input: &str) -> bril_rs::Program {
    use bril2json::parse_abstract_program_from_read;
    use bril_rs::load_program_from_read;
    let abs_program = parse_abstract_program_from_read(input.as_bytes(), true, false, None);
    let mut buf = Vec::new();
    serde_json::to_writer_pretty(&mut buf, &abs_program).unwrap();
    buf.push(b'\n');
    let json_str = String::from_utf8(buf).unwrap();
    load_program_from_read(json_str.as_bytes())
}

/// Visual representations of intermediate representations computed on a bril
/// program.
///
/// This struct is intended for use in debugging eggcc. SVG visualizations can
/// be opened directly in a web browser. Dot visualizations can be converted to
/// (e.g.) PNGs with utilities like graphviz, e.g.
/// `dot -Tpng input_cfg.dot -o input_cfg.png`.
#[allow(unused)]
pub struct DebugVisualizations {
    /// The raw CFG before any restructuring, in Dot format,
    /// for each function in the program.
    pub(crate) input_cfgs: HashMap<String, String>,
    /// The restructured CFG ready for RVSDG conversion, in Dot format.
    pub(crate) restructured_cfgs: HashMap<String, String>,
    /// The RVSDG, rendered in SVG format.
    pub(crate) rvsdg: String,
    /// The PEG, rendered in DOT format.
    pub(crate) peg: String,
}

impl DebugVisualizations {
    /// Compute visualizations for the given bril program.
    ///
    /// This function is intended for use in debugging.
    ///
    /// # Panics
    /// Any failures in the conversion to CFG or RVSDG will cause a panic.
    pub fn new(input: &str) -> DebugVisualizations {
        let program = parse_from_string(input);
        let cfg = Optimizer::program_to_cfg(&program);
        let rvsdg = Optimizer::program_to_rvsdg(&program).unwrap();

        let input_cfgs = cfg
            .functions
            .iter()
            .map(|f| (f.name.clone(), format!("{:#?}", Dot::new(&f.graph))))
            .collect();
        let restructured_cfgs = cfg
            .functions
            .iter()
            .map(|f| {
                let mut f = f.clone();
                f.restructure();
                (f.name.clone(), format!("{:#?}", Dot::new(&f.graph)))
            })
            .collect();

        DebugVisualizations {
            input_cfgs,
            restructured_cfgs,
            rvsdg: rvsdg.to_svg(),
            peg: rvsdg_to_peg(&rvsdg).graph(),
        }
    }

    /// Write the visualizations to output files in the given directory.
    /// If the directory does not exist, it creates it.
    /// If the directory contains any files whose names conflict with the
    /// output files, it replaces them.
    ///
    /// * The `input_cfg` field is written to `input_cfg.dot`.
    /// * The `restructured_cfg` field is written to `function_name` + `restructured.dot`.
    /// * The `rvsdg` field is written to `rvsdg.svg`.
    ///
    /// Like other utilities related to `DebugVisualizations`, this method is
    /// only intended for debugging eggcc.
    pub fn write_output(&self, path: PathBuf) -> io::Result<()> {
        use std::fs::File;
        use std::io::Write;

        // make the directory if it doesn't exist
        if !path.exists() {
            std::fs::create_dir_all(&path)?;
        }

        for (name, content) in
            self.input_cfgs
                .iter()
                .map(|(name, content)| (format!("{name}_cfg.dot"), content.clone()))
                .chain(self.restructured_cfgs.iter().map(|(name, content)| {
                    (format!("{name}_restructured_cfg.dot"), content.clone())
                }))
                .chain([("rvsdg.svg".to_string(), self.rvsdg.clone())])
                .chain([("peg.dot".to_string(), self.peg.clone())])
        {
            let mut output_path = path.clone();
            assert!(path.is_dir());
            output_path.push(name.clone());
            let mut file = File::create(output_path)?;
            file.write_all(content.as_bytes())?;

            // if it's a dot file, also write
            // an svg using the dot program:
            if name.ends_with(".dot") {
                let output_path = path.join(name.replace(".dot", ".svg"));
                self.run(
                    "dot",
                    [
                        "-Tsvg",
                        "-o",
                        &output_path.into_os_string().into_string().unwrap(),
                    ],
                    &content,
                )?;
            }
        }
        Ok(())
    }

    /// Invokes some program with the given arguments, piping the given input to the program.
    /// Returns an error if the program returns a non-zero exit code.
    /// Code adapted from https://github.com/egraphs-good/egg/blob/e7845c5ae34267256b544c8e6b5bc36d91d096d2/src/dot.rs#L127
    pub fn run<S1, S2, I>(&self, program: S1, args: I, input: &str) -> std::io::Result<()>
    where
        S1: AsRef<OsStr>,
        S2: AsRef<OsStr>,
        I: IntoIterator<Item = S2>,
    {
        use std::io::Write;
        use std::process::{Command, Stdio};
        let mut child = Command::new(program)
            .args(args)
            .stdin(Stdio::piped())
            .stdout(Stdio::null())
            .spawn()?;
        let stdin = child.stdin.as_mut().expect("Failed to open stdin");
        write!(stdin, "{}", input)?;
        match child.wait()?.code() {
            Some(0) => Ok(()),
            Some(e) => Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("dot program returned error code {}", e),
            )),
            None => Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                "dot program was killed by a signal",
            )),
        }
    }
}

#[derive(Clone, Copy)]
pub enum RunType {
    StructuredConversion,
    RvsdgConversion,
    NaiiveOptimization,
}

impl Display for RunType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            RunType::StructuredConversion => write!(f, "structured"),
            RunType::RvsdgConversion => write!(f, "rvsdg"),
            RunType::NaiiveOptimization => write!(f, "naiive"),
        }
    }
}

impl RunType {
    pub fn produces_bril(&self) -> bool {
        match self {
            RunType::StructuredConversion => false,
            RunType::RvsdgConversion => false,
            RunType::NaiiveOptimization => true,
        }
    }
}

#[derive(Clone)]
pub struct Run {
    pub path: PathBuf,
    pub test_type: RunType,
    // Also interpret the resulting program
    pub interp: bool,
}

#[derive(Clone)]
pub struct RunOutput {
    // a visualization of the result
    pub visualization: String,
    // a viable file extension for the visualization
    pub visualization_file_extension: String,
    // if the result was interpreted, the stdout of interpreting it
    pub result_interpreted: Option<String>,
    pub original_interpreted: String,
}

impl Run {
    pub fn all_configurations_for(path: PathBuf) -> Vec<Run> {
        let mut res = vec![];
        for test_type in [
            RunType::StructuredConversion,
            //RunType::RvsdgConversion,
            RunType::NaiiveOptimization,
        ] {
            let default = Run {
                path: path.clone(),
                test_type,
                interp: false,
            };
            res.push(default.clone());
            if test_type.produces_bril() {
                let interp = Run {
                    interp: true,
                    ..default
                };
                res.push(interp);
            }
        }
        res
    }

    pub fn name(&self) -> String {
        let mut name = self.path.file_stem().unwrap().to_str().unwrap().to_string();
        name = format!("{}-{}", name, self.test_type);
        if self.interp {
            name = format!("{}-interp", name);
        }
        name
    }

    pub fn run(&self) -> RunOutput {
        let program_read = std::fs::read_to_string(self.path.clone()).unwrap();
        let args = Optimizer::parse_bril_args(&program_read);
        let original_interpreted = Optimizer::interp(&program_read, args.clone(), None);
        let (visualization, visualization_file_extension) = match self.test_type {
            RunType::StructuredConversion => {
                let structured = Optimizer::parse_to_structured(&program_read).unwrap();
                (structured.to_string(), ".txt")
            }
            RunType::RvsdgConversion => {
                let parsed = Optimizer::parse_bril(&program_read).unwrap();
                let rvsdg = Optimizer::program_to_rvsdg(&parsed).unwrap();
                let svg = rvsdg.to_svg();
                (svg, ".svg")
            }
            RunType::NaiiveOptimization => {
                let parsed = Optimizer::parse_bril(&program_read).unwrap();

                let mut optimizer = Optimizer::default();
                let res = optimizer.optimize(&parsed).unwrap();

                (format!("{}", res), ".bril")
            }
        };

        let result_interpreted = if self.interp {
            Some(Optimizer::interp(&program_read, args.clone(), None))
        } else {
            None
        };

        RunOutput {
            visualization,
            visualization_file_extension: visualization_file_extension.to_string(),
            result_interpreted,
            original_interpreted,
        }
    }
}
