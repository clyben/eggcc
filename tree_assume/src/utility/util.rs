#[cfg(test)]
use crate::{egglog_test, interpreter::Value};

#[test]
fn test_list_util() -> crate::Result {
    let build = "
		(let list (Cons (Const (Int 0))
                  (Cons (Const (Int 1))
                  (Cons (Const (Int 2))
                  (Cons (Const (Int 3))
                  (Cons (Const (Int 4)) (Nil)))))))
		(let expr (Switch (Const (Int 1)) list))
	";
    let check = "
		(check (= (ListExpr-ith list 1) (Const (Int 1))))
        (check (= (ListExpr-ith list 4) (Const (Int 4))))
        (check (= (ListExpr-length list) 5))
	";
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
    )
}

#[test]
fn append_test() -> crate::Result {
    let build = "
        (let appended
            (Append
                (Cons (Const (Int 0)) (Cons (Const (Int 1)) (Nil)))
                (Const (Int 2))))
    ";

    let check = "
        (check (
            =
            (Cons (Const (Int 0)) (Cons (Const (Int 1)) (Cons (Const (Int 2)) (Nil))))
            appended
        ))
    ";
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
    )
}