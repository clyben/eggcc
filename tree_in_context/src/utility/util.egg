(function ListExpr-length (ListExpr) i64)
(function ListExpr-ith (ListExpr i64) Expr :unextractable)
(function ListExpr-suffix (ListExpr i64) ListExpr :unextractable)
(function Append (ListExpr Expr) ListExpr :unextractable)

(rule ((Switch pred branch)) ((union (ListExpr-suffix branch 0) branch)) :ruleset always-run)

(rule ((= (ListExpr-suffix top n) (Cons hd tl)))
    ((union (ListExpr-ith top n) hd)
     (union (ListExpr-suffix top (+ n 1)) tl)) :ruleset always-run)

(rule ((= (ListExpr-suffix list n) (Nil)))
    ((set (ListExpr-length list) n)) :ruleset always-run)

(rewrite (Append (Cons a b) e)
   (Cons a (Append b e))
   :ruleset always-run)
(rewrite (Append (Nil) e)
   (Cons e (Nil))
   :ruleset always-run)

(function tuple-ith (Expr i64) Expr :unextractable)
(function tuple-length (Expr) i64 :unextractable)

(rule ((HasType expr (TupleT tl))
       (= len (TypeList-length tl)))
      ((set (tuple-length expr) len)) :ruleset always-run)

(rule ((Single expr)) ((union (tuple-ith (Single expr) 0) expr)) :ruleset always-run)
(rule ((Concat ord expr1 expr2)
       (= ith (tuple-ith expr1 i)))
       ((union (tuple-ith (Concat ord expr1 expr2) i) ith)) :ruleset always-run)

(rule ((Concat ord expr1 expr2)
       (= len1 (tuple-length expr1))
       (= ith (tuple-ith expr2 i)))
       ((union (tuple-ith (Concat ord expr1 expr2) (+ len1 i)) ith)) :ruleset always-run)

;; This Expr size actually do the same thing as cost function, can be removed afterwards.
(function Expr-size (Expr) i64 :merge (min old new))
(function ListExpr-size (ListExpr) i64 :merge (min old new))

(rule ((= expr (Function name tyin tyout out)) 
       (= sum (Expr-size out))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Const n ty))) 
      ((set (Expr-size expr) 1))  :ruleset always-run)

(rule ((= expr (Bop op x y)) 
       (= sum (+ (Expr-size y) (Expr-size x)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Uop op x)) 
       (= sum (Expr-size x))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Get tup i)) 
       (= sum (Expr-size tup))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Concat order x y)) 
       (= sum (+ (Expr-size y) (Expr-size x)))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Single x)) 
       (= sum (Expr-size x))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Switch pred branches)) 
       (= sum (+ (ListExpr-size branches) (Expr-size pred)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (If pred then else)) 
       (= sum (+ (Expr-size else) (+ (Expr-size then) (Expr-size pred))))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (DoWhile in pred-and-output)) 
       (= sum (+ (Expr-size pred-and-output) (Expr-size in)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Let in out)) 
       (= sum (+ (Expr-size out) (Expr-size in)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Arg ty))) 
      ((set (Expr-size expr) 1)) :ruleset always-run)

(rule ((= expr (Call func arg)) 
       (= sum (Expr-size arg))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((Empty ty)) ((set (Expr-size (Empty ty)) 0))  :ruleset always-run)

(rule ((= expr (Cons hd tl)) 
       (= sum (+ (ListExpr-size tl) (Expr-size hd)))) 
      ((set (ListExpr-size expr) sum)) :ruleset always-run)

(rule ((Nil)) 
      ((set (ListExpr-size (Nil)) 0))  :ruleset always-run)

(rule ((= expr (Alloc e ty)) 
        (= sum (Expr-size e))) 
        ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (InContext assumption e)) 
       (= sum (Expr-size e))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)


(function tuple-of-get-arg (Type i64) Expr :unextractable)
(function tuple-of-get-arg-helper (Type i64 i64) Expr :unextractable)

(rewrite (tuple-of-get-arg ty n) (tuple-of-get-arg-helper ty n 0) :ruleset always-run)
(rewrite (tuple-of-get-arg-helper ty 1 x) (Single (Get (Arg ty) x)) :ruleset always-run)
(rule ((= f (tuple-of-get-arg-helper ty a b)) (> a 1))
      ((union f 
         (Concat par
               (Get (Arg ty) b) 
               (tuple-of-get-arg-helper ty (- a 1) (+ b 1))))) :ruleset always-run)