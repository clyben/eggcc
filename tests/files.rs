use eggcc::util::Run;
use insta::assert_snapshot;
use libtest_mimic::Trial;

fn generate_tests(glob: &str) -> Vec<Trial> {
    let mut trials = vec![];
    let mut mk_trial = |run: Run| {
        trials.push(Trial::test(run.name(), move || {
            let result = run.run();

            if let Some(interpreted) = result.result_interpreted {
                assert_eq!(result.original_interpreted, interpreted);
            } else {
                assert_snapshot!(run.name(), result.visualization);
            }
            Ok(())
        }))
    };

    for entry in glob::glob(glob).unwrap() {
        let f = entry.unwrap();
        let name = f
            .file_stem()
            .unwrap()
            .to_string_lossy()
            .replace(['.', '-', ' '], "_");

        // TODO optimizer doesn't support these yet
        let banned = ["queens_func", "unstructured", "implicit_return", "fib"];
        if banned.iter().any(|b| name.contains(b)) || f.to_str().unwrap().contains("failing") {
            continue;
        }

        for run in Run::all_configurations_for(f) {
            mk_trial(run);
        }
    }

    trials
}

fn main() {
    let args = libtest_mimic::Arguments::from_args();
    let tests = generate_tests("tests/**/*.bril");
    libtest_mimic::run(&args, tests).exit();
}
