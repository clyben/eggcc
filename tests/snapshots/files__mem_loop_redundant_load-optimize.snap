---
source: tests/files.rs
expression: visualization.result
---
@main(v0: bool) {
  v2: int = const 3;
  v5: ptr<int> = alloc v2;
  v8: int = const 1;
  v10: ptr<int> = ptradd v5 v8;
  v14: int = const 2;
  v16: ptr<int> = ptradd v5 v14;
  v19: ptr<int> = id v10;
  v20: ptr<int> = id v5;
  v21: ptr<int> = id v16;
  v22: bool = id v0;
.__24__:
  v19: ptr<int> = id v20;
  v20: ptr<int> = id v19;
  v21: ptr<int> = id v21;
  v22: bool = id v22;
  br v22 .__24__ .__32__;
.__32__:
  v33: int = const 10;
  v36: int = const 20;
  store v20 v33;
  store v21 v36;
  print v33;
  free v19;
}
